qiskit_dynamics/pulse/string_model_parser/string_model_parser.py

```diff
@@ -0,0 +1,305 @@
+        subsystem_list = subsystem_list.copy()
+        subsystem_list.sort()
+
+    # force keys in hamiltonian['qub'] to be ints
+    qub_dict = {int(key): val for key, val in hamiltonian_dict["qub"].items()}
+
+    subsystem_dims = {int(qubit): qub_dict[int(qubit)] for qubit in subsystem_list}
+
+    # Parse the Hamiltonian
+    system = legacy_parser(
+        operator_str=hamiltonian_dict["h_str"],
+        subsystem_dims=subsystem_dims,
+        subsystem_list=subsystem_list,
+    )
+
+    # Extract which channels are associated with which Hamiltonian terms.
+    # Assumes one channel appearing in each term appearing at the end.
+    channels = []
+    for _, ham_str in system:
+        chan_idx = None
+
+        for c in CHANNEL_CHARS:
+            # if c in ham_str, and all characters after are digits, treat
+            # as channel
+            if c in ham_str:
+                if all(a.isdigit() for a in ham_str[ham_str.index(c) + 1 :]):
+                    chan_idx = ham_str.index(c)
+                    break
+
+        if chan_idx is None:
+            channels.append(None)
+        else:
+            channels.append(ham_str[chan_idx:])
+
+    # evaluate coefficients
+    local_vars = {chan: 1.0 for chan in set(channels) if chan is not None}
+    local_vars.update(variables)
+
+    evaluated_ops = []
+    for op, coeff in system:
+        # pylint: disable=exec-used
+        exec("evaluated_coeff = %s" % coeff, globals(), local_vars)
+        evaluated_ops.append(local_vars["evaluated_coeff"] * op)
+
+    # merge terms based on channel
+    static_hamiltonian = None
+    hamiltonian_operators = []
+    reduced_channels = []
+
+    for channel, op in zip(channels, evaluated_ops):
+        # if None, add it to the static hamiltonian
+        if channel is None:
+            if static_hamiltonian is None:
+                static_hamiltonian = op
+            else:
+                static_hamiltonian += op
+        else:
+            channel = channel.lower()
+            if channel in reduced_channels:
+                hamiltonian_operators[reduced_channels.index(channel)] += op
+            else:
+                hamiltonian_operators.append(op)
+                reduced_channels.append(channel)
+
+    # sort channels/operators according to channel ordering
+    if len(reduced_channels) > 0:
+        reduced_channels, hamiltonian_operators = zip(
+            *sorted(zip(reduced_channels, hamiltonian_operators))
+        )
+
+    return static_hamiltonian, list(hamiltonian_operators), list(reduced_channels), subsystem_dims
+
+
+def hamiltonian_pre_parse_exceptions(hamiltonian_dict: dict):
+    """Raises exceptions for improperly formatted or unsupported elements of
+    hamiltonian dict specification.
+
+    Parameters:
+        hamiltonian_dict: Dictionary specification of hamiltonian.
+    Returns:
+    Raises:
+        QiskitError: If some part of the Hamiltonian dictionary is unsupported or invalid.
+    """
+
```

Comment:
Some other ideas for preparsing if we want to (maybe unnecessary)

- check if variables exist which aren't defined (probably handled by exceptions from the evaluation).
- Check for non-alphanumeric characters that don't have predefined uses (unless those are fine as variable names)

Reply:
These are good ideas, though interestingly these problems are actually automatically handled gracefully by virtue of how these expressions are evaluated. Since `exec` is called on the coefficient strings, if a variable isn't defined a `NameError` is raised for that variable name. I think the same applies to the second point, and the legacy parser will handle undefined operator names (via calls to `operator_from_string`).

Reply:
great this was my hope